### PR TITLE
raftstore: support dynamically scale snap-generator-pool-size

### LINKED ISSUE #14325: Make snapshot thread pool size related parameters dynamic configurable
## Development Task
Sometimes customers want to scale-out or scale-in nodes ASAP. And today the default 2 threads become bottleneck even if customers lift all other related parameters (in both PD and TiKV), like below
pd-ctl store limit all 200
pd-ctl config set leader-schedule-limit 2147483647
pd-ctl config set region-schedule-limit 2147483647
pd-ctl config set max-pending-peer-count 2147483647
pd-ctl config set max-snapshot-count 2147483647
pd-ctl config set replica-schedule-limit 2147483647

<delete store 3009 command issued>
pd-ctl store limit 3009 2147483647
TiKV:
  SET CONFIG tikv `server.snap-max-write-bytes-per-sec` = '300MB'    

And to improve the scale-out or scale-in performance, they have to tune the parameter snap-generator-pool-size, but that requires tikv restart, which is inconvenient and sometimes impractical.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #14325

What's Changed:

```commit-message
support dynamically scale snap-generator-pool-size
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/worker/region.rs [hunk 1]
```diff
@@ -844,7 +851,7 @@ where
                     start: UnixSecs::now(),
                 };
                 let scheduled_time = Instant::now_coarse();
-                self.pool.spawn(async move {
+                let _ = self.pool.spawn(async move {
```
Comment (bufferflies) [inline components/raftstore/src/store/worker/region.rs:854]: Do you needs to handle the error?

File: components/raftstore/src/store/worker/region.rs [hunk 2]
```diff
@@ -407,12 +405,21 @@ where
             coprocessor_host,
             router,
             pd_client,
-            pool: Builder::new(thd_name!("snap-generator"))
-                .max_thread_count(cfg.value().snap_generator_pool_size)
+            pool: YatpPoolBuilder::new(DefaultTicker::default())
+                .name_prefix("snap-gen")
```
Comment (bufferflies) [inline components/raftstore/src/store/worker/region.rs:409]: why change the thread name?

File: components/raftstore/src/store/worker/region.rs [hunk 3]
```diff
@@ -54,6 +51,7 @@ use crate::{
 };
 
 const CLEANUP_MAX_REGION_COUNT: usize = 64;
+const SNAP_GENERATOR_MAX_POOL_SIZE: usize = 8;
```
Comment (bufferflies) [inline components/raftstore/src/store/worker/region.rs:54]: I think it maybe too small.

File: components/raftstore/src/store/worker/region.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/region.rs:409]: It's for testing. I will change it back.

File: components/raftstore/src/store/worker/region.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/region.rs]: Doubled it.

File: tests/integrations/raftstore/test_scale_pool.rs [hunk 4]
```diff
@@ -556,3 +559,124 @@ fn test_adjust_hight_priority_background_threads() {
...
+            break;
+        }
+    }
+    assert!(engine.get_value(b"zkey0001").unwrap().is_none());
+
+    fail::remove("before_region_gen_snap");
+}
+
+#[test]
+fn test_decrease_snap_generator_pool_size() {
```
Comment (overvenus) [inline tests/integrations/raftstore/test_scale_pool.rs:637]: Does it support raftstore-v2?

File: components/raftstore/src/store/worker/region.rs [hunk 5]
```diff
@@ -409,12 +410,16 @@ where
                 .thread_count(
                     1,
                     cfg.value().snap_generator_pool_size,
```
Comment (overvenus) [inline components/raftstore/src/store/worker/region.rs:412]: Can it be zero?

File: tests/integrations/raftstore/test_scale_pool.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_scale_pool.rs:637]: v2 does not have snap generator worker. v2 uses async read worker for checkpoint which is a cheap work.

File: components/raftstore/src/store/worker/region.rs (hunk 5 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/region.rs:412]: no

File: tests/integrations/raftstore/test_scale_pool.rs [hunk 6]
```diff
@@ -556,3 +559,124 @@ fn test_adjust_hight_priority_background_threads() {
...
+    fail::cfg("before_region_gen_snap", "1*pause").unwrap();
+
+    {
+        let sim = cluster.sim.rl();
+        let cfg_controller = sim.get_cfg_controller(1).unwrap();
+        let change = {
+            let mut change = HashMap::new();
+            change.insert(
+                "raftstore.snap-generator-pool-size".to_owned(),
+                "1".to_owned(),
```
Comment (overvenus) [inline tests/integrations/raftstore/test_scale_pool.rs:663]: Could you test "0" too?

**Review by overvenus — APPROVED** [review]

**Review by bufferflies — APPROVED** [review]

**Pull request merged**